bluebutton-data-pipeline-rif-extract/src/main/java/gov/hhs/cms/bluebutton/datapipeline/rif/extract/s3/DataSetMonitorWorker.java

```diff
@@ -0,0 +1,283 @@
+							&& (dataSetTimestamp.compareTo(parseDataSetTimestamp(manifestToProcessKey)) < 0))
+						manifestToProcessKey = key;
+				}
+			}
+
+			objectListing = s3Client.listNextBatchOfObjects(objectListing);
+		} while (objectListing.isTruncated());
+
+		// If no manifest was found, we're done (until next time).
+		if (manifestToProcessKey == null)
+			return;
+
+		// We've found the oldest manifest. Now go download and parse it.
+		DataSetManifest dataSetManifest = readManifest(manifestToProcessKey);
+
+		/*
+		 * We've got a dataset to process. However, it might still be uploading
+		 * to S3, so we need to wait for that to complete before we start
+		 * processing it.
+		 */
+		boolean dataSetComplete = false;
+		do {
+			if (dataSetIsAvailable(dataSetManifest))
+				dataSetComplete = true;
+
+			/*
+			 * We'll loop forever, but it's prudent to pause between each
+			 * iteration.
+			 */
+			try {
+				Thread.sleep(1000 * 1);
+			} catch (InterruptedException e) {
+				/*
+				 * Many Java applications use InterruptedExceptions to signal
+				 * that a thread should stop what it's doing ASAP. This app
+				 * doesn't, so this is unexpected, and accordingly, we don't
+				 * know what to do. Safest bet is to blow up.
+				 */
+				throw new RuntimeException(e);
+			}
+		} while (!dataSetComplete);
+
+		/*
+		 * Huzzah! We've got a data set to process and we've verified it's all
+		 * there and ready to go. Now we can hand that off to the
+		 * DataSetProcessor, to do the *real* work of actually processing that
+		 * data set. It's important that we block until it's completed, in order
+		 * to ensure that we don't end up processing multiple data sets in
+		 * parallel (which would lead to data consistency problems).
+		 */
+		Set<RifFile> rifFiles = dataSetManifest.getEntries().stream()
+				.map(e -> String.format("%s/%s", DateTimeFormatter.ISO_INSTANT.format(dataSetManifest.getTimestamp()),
+						e.getName()))
+				.map(k -> new S3RifFile(s3Client, new GetObjectRequest(bucketName, k))).collect(Collectors.toSet());
+		RifFilesEvent rifFilesEvent = new RifFilesEvent(dataSetManifest.getTimestamp(), rifFiles);
+		dataSetProcessor.process(rifFilesEvent);
+
+		/*
+		 * Now that the data set has been processed, we need to remove it. If we
+		 * don't, we'll just end up reprocessing it again the next time the
+		 * worker runs.
+		 */
+		DeleteObjectsRequest deleteObjectsRequest = new DeleteObjectsRequest(bucketName);
+		deleteObjectsRequest.setKeys(dataSetManifest.getEntries()
+				.stream().map(e -> String.format("%s/%s",
+						DateTimeFormatter.ISO_INSTANT.format(dataSetManifest.getTimestamp()), e.getName()))
+				.map(k -> new KeyVersion(k)).collect(Collectors.toList()));
+		deleteObjectsRequest.getKeys().add(new KeyVersion(String.format("%s/manifest.xml",
+				DateTimeFormatter.ISO_INSTANT.format(dataSetManifest.getTimestamp()))));
+		s3Client.deleteObjects(deleteObjectsRequest);
```

Comment:
I don't know about special error handling, but yeah, we need error handling in general. Just not addressing that yet.